### ISSUE TITLE
wrong answers with 0.3.7 on POWER8 and POWER9

### ISSUE THREAD

**Issue by cparrott73** [xianyi/OpenBLAS#2233]
Not sure if you guys were already aware of this, but I wanted to pass this one along, just in case.

I was trying out the OpenBLAS 0.3.7 release on our POWER8 ("garrison") and POWER9 ("newell") OpenPOWER systems here this week.  I used GCC 9.2.0, although I have also observed similar behavior with GCC 8.3.0, which leads me to believe that this issue is not related to a particular release of GCC.

The library compiles fine, but several tests are failing on both platforms with accuracy errors:

```
 CGEMV  PASSED THE TESTS OF ERROR-EXITS

 ******* FATAL ERROR - COMPUTED RESULT IS LESS THAN HALF ACCURATE *******
                       EXPECTED RESULT                    COMPUTED RESULT
       1  (  -0.102385    ,   0.226428    )  (    0.00000    ,    0.00000    )
       2  (   0.108077    ,   0.805703E-01)  (  -0.102385    ,   0.226428    )
       3  (   0.992165E-02,  -0.840403E-01)  (   0.108077    ,   0.805703E-01)
       4  (   0.192725    ,   0.176668    )  (   0.992165E-02,  -0.840403E-01)
 ******* CGEMV  FAILED ON CALL NUMBER:
    512: CGEMV ('N',  4,  2,( 1.0, 0.0), A,  5, X,-1,( 0.0, 0.0), Y, 1)         .
```

I see similar issues with CGBMV, CHEMV, CHPMV, CTRMV, CTPMV, CTRSV, CTPSV, CGERC, CGERU, CHER, CHPR, CHER2, and CHPR2.

The cblas_cgemv test also fails:

```
 ******* FATAL ERROR - COMPUTED RESULT IS LESS THAN HALF ACCURATE *******
                       EXPECTED RESULT                    COMPUTED RESULT
       1  (  -0.342872    ,   0.714919E-02)  (  -0.342872    ,   0.714919E-02)
       2  (   0.233743    ,  -0.822020E-01)  (   0.233743    ,  -0.822020E-01)
       3  (   0.376182E-01,   0.161061    )  (   0.376182E-01,   0.161061    )
       4  (   0.144652    ,   0.137748    )  (    0.00000    ,    0.00000    )
 ******* cblas_cgemv  FAILED ON CALL NUMBER:
    440: cblas_cgemv (  CblasNoTrans,  4,  2,( 1.0, 0.0), A,
            5, X, 1,( 0.0, 0.0), Y, 1) .
 ******* cblas_cgemv  FAILED ON CALL NUMBER:
      4: cblas_cgemv (  CblasNoTrans,  2,  1,( 0.0, 0.0), A,
            3, X, 1,( 0.0, 0.0), Y, 1) .

 ******* FATAL ERROR - TESTS ABANDONED *******
```

I built this using the following command:

```
make CC=gcc FC=gfortran USE_OPENMP=1 NUM_THREADS=512 DYNAMIC_ARCH=1 all
```

I also tried this with USE_OPENMP=0, but still got the same failures.

Environment details:

CPU: POWER8, POWER9
OS: RHEL 7.5
Compiler: GCC 9.2.0, GCC 8.3.0 (both built internally from sources)

Thanks in advance.

**Comment by martin-frbg** [comment]
This would appear to point to a problem in #1994 - in that case the errors should go away when you switch back kernel/power/KERNEL.POWER8 (and KERNEL.POWER9) to define `CGEMVNKERNEL=../arm/zgemv_n.c` and `CGEMVTKERNEL=../arm/zgemv_t.c` instead of the new, local cgemv_n.c/cgemv_t.c

**Comment by cparrott73** [comment]
Thanks, Martin.  I tried reverting the kernels for CGEMV as you suggested, and indeed, the tests now pass:

```
 CGEMV  PASSED THE TESTS OF ERROR-EXITS

 CGEMV  PASSED THE COMPUTATIONAL TESTS (  4324 CALLS)
```

I presume I would need to do a similar reversion for CGBMV and the other failures?

Thanks so much.

+chris

**Comment by martin-frbg** [comment]
Actually I would hope that these errors will be gone as well, once you do a full recompile. The "banded matrix" etc. variants should all use essentially the same "general" CGEMV microkernel

**Comment by cparrott73** [comment]
You're right.  Not sure what happened before, but I just cleaned everything and then commented out those two lines in for CGEMVNKERNEL and CGEMVTKERNEL in KERNEL.POWER8- essentially reverting to them how they were before - and everything passes now after I rebuilt.

**Comment by cparrott73** [comment]
Correction - I did end up having to revert several other files, too.  Here are the lines I ended up commenting out in KERNEL.POWER8:

```
#SAXPYKERNEL  = saxpy.c
#CAXPYKERNEL  = caxpy.c

#CDOTKERNEL = cdot.c

#SGEMVNKERNEL = sgemv_n.c
#CGEMVNKERNEL = cgemv_n.c

#SGEMVTKERNEL = sgemv_t.c
#CGEMVTKERNEL = cgemv_t.c
```

Uncommenting any one of these seems to be sufficient to cause test failures.

**Comment by martin-frbg** [comment]
Our Ci test is passing (e.g. https://travis-ci.org/xianyi/OpenBLAS/jobs/573911250), but it is using an older (5.2) release of gcc. In the past, misbehaviour with gcc 8 was usually a sign of mistakes in our inline assembly, brought to light by more aggressive register reuse. However the new cgemv files do not even use inline assembly directly.  @quickwritereader ?

**Comment by cparrott73** [comment]
Interesting.  I figured I should try a newer GCC for POWER9 support.  GCC 5.x doesn't seem to support the -mcpu=power9 argument, at least when I tried it here.

I tried GCC 6.5.0 here instead, which does appear to support -mcpu=power9.  Indeed, I did not get the test failures this time.  So I would concur with your assessment that newer GCC versions must be doing something involving register selection that is messing with the inline assembly in these routines.

It's not an issue for us to use GCC 6.5.0 to build OpenBLAS here, for the time being.  Thanks.

**Comment by martin-frbg** [comment]
I have now requested&obtained an account on the unicamp.br minicloud - initial POWER8 compile tests there show that building with gcc/gfortran 7.3.1 works flawlessly, but a 8.2.1 build shows failures in a lot of the single-precision complex tests. (CGEMV, CGBMV, CHBMV, CHPMV, CTRMV, CTPMV, CTRSV and CTPSV are flagged as "less than half accurate", often with the deviating values in the 1e10 range . Others like CGERC and CGERU get "parameter 8 was changed incorrectly". Stack corruption ?)

**Comment by quickwritereader** [comment]
I just saw. could you elaborate on what I should check

**Comment by quickwritereader** [comment]
Passed for me using these: 
  Operating System: CentOS Linux 7 (AltArch):

gcc (GCC) 8.3.1 20190304 (Advance-Toolchain-at12.0) [revision 269374]
Copyright (C) 2018 Free Software Foundation, Inc. 

----------------------  

Operating System: Ubuntu 18.04
gcc (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
Copyright (C) 2017 Free Software Foundation, Inc. E.

**Comment by quickwritereader** [comment]
Yes,  failed codes by op are written with vector C (So, there are not any inline assembly codes ). So it can be reviewed easily. Sometimes, it is hard to spot your own problems.

**Comment by martin-frbg** [comment]
Most if not all of the "LESS THAN HALF ACCURATE" errors from the lapack test appear to be related to the new CAXPY. Trying to narrow it down further... my tests are run on openSUSE/ppc64le on POWER8 hardware as provided by unicamp. I have not yet tried to build my own gcc9 there

**Comment by quickwritereader** [comment]
@martin-frbg if it is possible could you re-check caxpy using BLAS_TESTER ,too
frankly, I don't know what can go wrong. if errors persist I will try to compile those version of gcc to test at the week-end
./xcl1blastst -N 1 1331 1

**Comment by martin-frbg** [comment]
As individual tests for some of these functions still work, I suspect we are seeing some unexpected side effect of (using) the codes from altivec.h. Not sure why this happens only with the newer gcc,
would it perhaps enforce a requirement for the -maltivec option (which does not appear to be set in Makefile.power) ? (The xcl1blastst passes with both old and new versions of the caxpy kernel)

**Comment by quickwritereader** [comment]
well, I think that gcc somehow messing ABI while translating if it gets parameters changed error. Such problems arise when writing assembly code by hand and when you improperly use registers.

**Comment by quickwritereader** [comment]
ok so why I'm getting this (zaxpy.c instead of  caxpy.c ):
gcc -Ofast -mcpu=power9 -mtune=power9 -mvsx -malign-power -DUSE_OPENMP -fno-fast-math -fopenmp -DMAX_STACK_ALLOC=2048 -fopenmp -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DSMP_SERVER -DUSE_OPENMP -DNO_WARMUP -DMAX_CPU_NUMBER=128 -DMAX_PARALLEL_NUMBER=1 -DVERSION=\"0.3.6.dev\" -DASMNAME=cblas_caxpy -DASMFNAME=cblas_caxpy_ -DNAME=cblas_caxpy_ -DCNAME=cblas_caxpy -DCHAR_NAME=\"cblas_caxpy_\" -DCHAR_CNAME=\"cblas_caxpy\" -DNO_AFFINITY -I.. -I. -UDOUBLE  -DCOMPLEX -DCBLAS **-c zaxpy.c** -o cblas_caxpy.o

**Comment by martin-frbg** [comment]
My feeling as well. The strange thing is that this does not happen with the older codebase that was using much more handcrafted assembly. So either there is some magic setup (or cleanup) intrinsic missing in your new code (and I see now that adding back the -maltivec that is commented out in Makefile.power does not change anything), or it is a victim of ongoing register misuse despite my attempts to fix the old code (#1263)

**Comment by martin-frbg** [comment]
gcc8-built library works at -O0 (passing all BLAS-Tester checks as well), but fails already at -O1, maybe narrowing down which of the added options exposes the problem will provide a hint. (Reminds me of the recent issue on x86_64 with operands declared as input-only despite being (ab)used as loop counters internally.)

**Comment by quickwritereader** [comment]
I tested using gcc-8.3  on ubuntu 
It passes  cgemv.

**Comment by quickwritereader** [comment]
ubuntu@openblas-p9:~/qwr/OpenBLAS$ gcc --version
gcc (Ubuntu 8.3.0-6ubuntu1~18.04.1) 8.3.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

ubuntu@openblas-p9:~/qwr/OpenBLAS$ gfortran --version
GNU Fortran (Ubuntu 8.3.0-6ubuntu1~18.04.1) 8.3.0

**Comment by martin-frbg** [comment]
Strangely, gcc 8.2.1 -O0 plus all the documented options from -O1 passed just now as well. Possibly a Heisenbug, or related to running a somewhat older version of binutils (In my case, 8.2.1 is supplied as a separate gcc-8 package where the system default is 7.3.1,  cparrott73 used self-built gcc 8.x and 9.x on a RHEL7.5 that seems to come with gcc 6.5 or older by default)

**Comment by quickwritereader** [comment]
Maybe issue happens when mixing old  gfortran with new gcc.

**Comment by martin-frbg** [comment]
Not likely (in my case at least), and IIRC this leads to crashes everywhere, not just wrong results from a select few single-precision complex functions.

**Comment by quickwritereader** [comment]
I've never used -fsanitize of gcc. it seems we can check with it for errors.
As It passed on my pc then there should not be a computational error but   ,probably , ub or similiar thing.

**Comment by martin-frbg** [comment]
Hmm. I have now created an Ubuntu 18.04 instance in the unicamp cloud, and get the same 22 counts of "HALF ACCURATE" in the make output there (gcc-8.3.0, gfortran-8.3.0). Where our systems will/may differ - (a) unicamp.br provides virtual instances running in qemu (though on actual ppc hardware) and (b) I created my instance to have two cpu cores.

**Comment by quickwritereader** [comment]
are you on slack team? I'd like to get access to that pc.

**Comment by martin-frbg** [comment]
Seems to me their Terms of Service do not allow sharing accounts unfortunately. (But getting one was easy and quick.) Anything specific you want me to check ?

**Comment by quickwritereader** [comment]
@martin-frbg Thanks.
 I will recheck it at the weekend. Frankly, I did not understand what might go wrong. I will try to use sanitize thing to see if there are problems inside code.
I checked on two computers with gcc 8.3 and gcc7. and it passed.
I was using simple make call

**Comment by martin-frbg** [comment]
I still do not understand it either - and it does not appear to depend on the number of cpus. Seeing it happen on 18.04 disproves my theory that older binutils could have anything to do with it - and that it works for you (assuming that you still have access to actual power8 hardware) makes it all the more strange.

**Comment by quickwritereader** [comment]
I was testing on  power9 CPUs

**Comment by quickwritereader** [comment]
just tested on power8 but with gcc-7 and it passed. But as I see gcc-7 was not an issue

**Comment by martin-frbg** [comment]
Indeed a gcc 8.3 build works flawlessly on POWER9 - and the number of different kernel files is quite small and does not include any of the files cparrott73 mentioned above. (Actually what stands out is the cgemm_tcopy_8_power8.S, I'll try replacing it with the generic C code used on POWER9 and see if that helps.)
Edited to add: the (only) effect of this was to turn three previous "parameter changed incorrectly" errors into "less than half accurate" (CGERC, CGERU and CHER).

**Comment by martin-frbg** [comment]
Going back to the generic C kernels for CTRMM and CGEMM makes everything work again - pity they can't be swapped out separately (without extra work, as there are no generic 8x4 replacements for either). No idea yet if the original code is to blame, or my later attempt to save&restore the VSX registers.

**Comment by quickwritereader** [comment]
changing them affects blas1 routines?

**Comment by martin-frbg** [comment]
I know it should not, but if we are working from the assumption that "something" is trashing registers or stack everything may be possible...

**Comment by quickwritereader** [comment]
your save/restore codes seems fine ~~.removal  which of them fixed it ? cgemm or ctrmm?~~ Nevermind I forgot that both are used inside trmm.   
What can I say about cgemm. It is using internal stack copy for matrixe blocks which I removed while porting power9.

[cgemm](https://github.com/xianyi/OpenBLAS/blob/develop/kernel/power/cgemm_kernel_8x4_power8.S)  
```
 #ifdef __64BIT__

#define STACKSIZE 32196

---- 
	addi	SP, SP, -STACKSIZE

	addi	SP, SP, -STACKSIZE

	addi	SP, SP, -STACKSIZE

	addi	SP, SP, -STACKSIZE
```

**Comment by martin-frbg** [comment]
For want of a better idea I tried using your power9 kernel, figuring that it would lead to an illegal instruction error at worst. Now I just wish I hadn't, as I got a whole lot of "ACCURACY" errors from blas2 functions before the SIGILL...

**Comment by martin-frbg** [comment]
Trying with clang now (needs a trivial fix in Makefile.power as clang does not understand -malign-power) - and it complains:
```
../kernel/power/dgemv_t.c:358:28: error: invalid lvalue in asm output
            : [memy] "+m" (*(const double (*)[8])y),
                           ^~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```
and
```
../kernel/power/dgemv_t.c:372:27: error: dereference of pointer to incomplete type 'const double []'
            [mem_ap] "m" (*(const double (*)[]) ap),
                          ^
```

**Comment by quickwritereader** [comment]
try if it works without converting just *ap *y

**Comment by quickwritereader** [comment]
I think with clang it will get messier as building with clang never was a goal.

Lets recap :  
* problem could be related to the computational inaccuracy
* problem related to ABI because of "parameter changed incorrectly"
* undefined behavior or   uninitialized variables

the problems pop while using >gcc-8 on power8. 
Reverting blas1 codes to generic c fixes.   
Reverting cgemm kernels also fixes the problem accuracy problem  

I think I will investigate firstly ABI error finding what changes parameter 8. This way I could see what register was not saved restored appropriately inside binary.

**Comment by quickwritereader** [comment]
> Trying with clang now (needs a trivial fix in Makefile.power as clang does not understand -malign-power) - and it complains:
> ../kernel/power/dgemv_t.c:358:28: error: invalid lvalue in asm output
>             : [memy] "+m" (*(const double (*)[8])y),
>                            ^~~~~~~~~~~~~~~~~~~~~~~
> 1 error generated.
> 
> and
> ../kernel/power/dgemv_t.c:372:27: error: dereference of pointer to incomplete type 'const double []'
>             [mem_ap] "m" (*(const double (*)[]) ap),
>                           ^

Other option would be just #undef -ing "#define HAVE_KERNEL4x8_ASM 1" . it should fallback to vector c code.

**Comment by martin-frbg** [comment]
clang8 build had ACCURACY errors for SGEMV, SSYMV,SGER, DGEMV in addition to CGEMV, CGBMV, CHEMV, CHPMV, CTRMV, CTRSV, CTPSV, CGERC, CGERU, CHER, CHPR, CHPR2 and also reported changed parameters for CGERC, CGERU, CHER, CHER2. Probably best to ignore this for now. (At -O1 instead of the default -Ofast, only the SGEMV error remains)

**Comment by quickwritereader** [comment]
martin could you supply
```
export OMP_NUM_THREADS=1
./xccblat2 <cin2
```
for me  final results are shifted  


                       EXPECTED RESULT                    COMPUTED RESULT
       1  (  -0.342872    ,   0.714919E-02)   **2)**  (   0.233743    ,  -0.822020E-01)
       2  (   0.233743    ,  -0.822020E-01)  **3)** (   0.376182E-01,   0.161061    )
       3  (   0.376182E-01,   0.161061    )  **4)** (   0.144652    ,   0.137748    )
       4  (   0.144652    ,   0.137748    )  (    0.00000    ,    0.00000    )

**Comment by quickwritereader** [comment]
I remembered that I saw such behaviour once. Probably,  lost that after toolchain changes

**Comment by martin-frbg** [comment]
Same values (with gcc, and all changes reverted except the undefining of the dgemv_t 4x8_ASM):
```
******* FATAL ERROR - COMPUTED RESULT IS LESS THAN HALF ACCURATE *******
                       EXPECTED RESULT                    COMPUTED RESULT
       1  (  -0.342872    ,   0.714919E-02)  (   0.233743    ,  -0.822020E-01)
       2  (   0.233743    ,  -0.822020E-01)  (   0.376182E-01,   0.161061    )
       3  (   0.376182E-01,   0.161061    )  (   0.144652    ,   0.137748    )
       4  (   0.144652    ,   0.137748    )  (    0.00000    ,    0.00000    )
 ******* cblas_cgemv  FAILED ON CALL NUMBER:
    440: cblas_cgemv (  CblasNoTrans,  4,  2,( 1.0, 0.0), A,
            5, X, 1,( 0.0, 0.0), Y, 1) .
 ******* cblas_cgemv  FAILED ON CALL NUMBER:
      4: cblas_cgemv (  CblasNoTrans,  2,  1,( 0.0, 0.0), A,
            3, X, 1,( 0.0, 0.0), Y, 1) .
```
Took me a moment to realize what you meant by "shifted", made me wonder if you could do bit shifts on the internal representation of floats in your head...
Edit: with the other errors it is less "obvious" - mostly individual numbers in the middle of a column
that are wrong, or only the last line wrong (CHER,CHPR). Trying `#pragma GCC optimize "O0"` in cgemv_n.c and cgemv_t.c next...

**Comment by martin-frbg** [comment]
Adding `#pragma GCC optimize "O0"` to cgemv_n.c, cgemv_t.c, caxpy.c and cdot.c appears to be a (temporary) solution, at least as far as the built-in tests go. (No idea what that means - though the common factor of these four files is that they use vector intrinsics. Perhaps it is possible to make unsafe assumptions about register (re)usage there as well, with the potential to hit seemingly unrelated functions that use the same registers later ?)

**Comment by quickwritereader** [comment]
for now I am suspicious of  swap_mask . I wrote simple code and it did not load fully

**Comment by quickwritereader** [comment]
~~Possible solution. 
`static const char __attribute__((aligned(16)))  swap_mask_arr[]={,,}`
with pure assembly I would use constant loading, anyway~~
Alignment did not work. I will recheck ,later.

**Comment by quickwritereader** [comment]
So far the problem in memory alignment. somehow gcc-8 chooses old loads which expect aligned memory .

**Comment by quickwritereader** [comment]
Ok. So far I cant understand why it chooses   load 16byte-aligned lvx  over (afaik 4byte or unaligned) lxv
Here is assemblies
gcc-8
```
   │0x10007390 <cgemv_n+2336>       lvx     v1,r5,r9                                                                                                                       │
   │0x10007394 <cgemv_n+2340>       lvx     v0,r30,r9                                                                                                                      │
   │0x10007398 <cgemv_n+2344>       addi    r8,r8,2                                                                                                                        │
   │0x1000739c <cgemv_n+2348>       lvx     v12,r7,r9                                                                                                                      │
   │0x100073a0 <cgemv_n+2352>       lvx     v13,r6,r9                                                                                                                      │
   │0x100073a4 <cgemv_n+2356>       cmpd    cr7,r3,r8                                                                                                                      │
   │0x100073a8 <cgemv_n+2360>       lvx     v5,r31,r9                                                                                                                      │
   │0x100073ac <cgemv_n+2364>       lvx     v6,r10,r9                                                                                                                      │
   │0x100073b0 <cgemv_n+2368>       xvmulsp vs10,vs12,vs33                                                                                                                 │
   │0x100073b4 <cgemv_n+2372>       xvmulsp vs11,vs12,vs32                                                                                                                 │
   │0x100073b8 <cgemv_n+2376>       vperm   v9,v12,v12,v11                                                                                                                 │
   │0x100073bc <cgemv_n+2380>       vperm   v10,v13,v13,v11                                                                                                                │
   │0x100073c0 <cgemv_n+2384>       vperm   v1,v1,v1,v11                                                                                                                   │
   │0x100073c4 <cgemv_n+2388>       vperm   v0,v0,v0,v11                                                                                                                   │
   │0x100073c8 <cgemv_n+2392>       xvmaddmsp vs44,vs0,vs10                                                                                                                │
   │0x100073cc <cgemv_n+2396>       xvmaddmsp vs45,vs0,vs11    

```
gcc-7
```
   lxvd2x  vs43,r7,r9                                                                                                                     │
   │0x10006bf4 <cgemv_n+1860>       lxvd2x  vs33,r7,r10                                                                                                                    │
  >│0x10006bf8 <cgemv_n+1864>       lxvd2x  vs44,r8,r9                                                                                                                     │
   │0x10006bfc <cgemv_n+1868>       lxvd2x  vs32,r8,r10                                                                                                                    │
   │0x10006c00 <cgemv_n+1872>       lxvd2x  vs0,r31,r10                                                                                                                    │
   │0x10006c04 <cgemv_n+1876>       lxvd2x  vs10,r31,r9                                                                                                                    │
   │0x10006c08 <cgemv_n+1880>       xxswapd vs43,vs43                                                                                                                      │
   │0x10006c0c <cgemv_n+1884>       xxswapd vs33,vs33                                                                                                                      │
   │0x10006c10 <cgemv_n+1888>       xxswapd vs44,vs44                                                                                                                      │
   │0x10006c14 <cgemv_n+1892>       xxswapd vs32,vs32                                                                                                                      │
   │0x10006c18 <cgemv_n+1896>       xvmulsp vs11,vs7,vs33                                                                                                                  │
   │0x10006c1c <cgemv_n+1900>       xvmulsp vs9,vs7,vs43                                                                                                                   │
   │0x10006c20 <cgemv_n+1904>       vperm   v10,v12,v12,v9                                                                                                                 │
   │0x10006c24 <cgemv_n+1908>       vperm   v13,v0,v0,v9                                                                                                                   │
   │0x10006c28 <cgemv_n+1912>       vperm   v11,v11,v11,v9
```

**Comment by quickwritereader** [comment]
gcc-8 doing good work getting rid of unnecessary xxswapd but using l**v**x (MEM & 0xFFFFFF...F0) instructor for loads mostly. but gcc-7 is using  l**x**v (x as the second letter) most times avoiding  l**v**x  (v in the middle)

Here is a simple command to check assemblies

```
cd kernel/ && gcc-7 -S -Ofast -mcpu=power8 -mtune=power8 -mvsx -malign-power -DUSE_OPENMP -fno-fast-math -fopenmp -g -DMAX_STACK_ALLOC=2048 -fopenmp -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DNO_LAPACK -DNO_LAPACKE -DSMP_SERVER -DUSE_OPENMP -DNO_WARMUP -DMAX_CPU_NUMBER=2 -DMAX_PARALLEL_NUMBER=1 -DVERSION=\"0.3.8.dev\" -DASMNAME=cgemv_s -DASMFNAME=cgemv_s_ -DNAME=cgemv_s_ -DCNAME=cgemv_s -DCHAR_NAME=\"cgemv_s_\" -DCHAR_CNAME=\"cgemv_s\" -DNO_AFFINITY -I.. -UDOUBLE  -DCOMPLEX -UDOUBLE -DCOMPLEX -UTRANS -DCONJ -DXCONJ ../kernel/power/cgemv_n.c -o ../cgcc-7_cgemv.S && cd ..

cd kernel/ && gcc-8 -S -Ofast -mcpu=power8 -mtune=power8 -mvsx -malign-power -DUSE_OPENMP -fno-fast-math -fopenmp -g -DMAX_STACK_ALLOC=2048 -fopenmp -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DNO_LAPACK -DNO_LAPACKE -DSMP_SERVER -DUSE_OPENMP -DNO_WARMUP -DMAX_CPU_NUMBER=2 -DMAX_PARALLEL_NUMBER=1 -DVERSION=\"0.3.8.dev\" -DASMNAME=cgemv_s -DASMFNAME=cgemv_s_ -DNAME=cgemv_s_ -DCNAME=cgemv_s -DCHAR_NAME=\"cgemv_s_\" -DCHAR_CNAME=\"cgemv_s\" -DNO_AFFINITY -I.. -UDOUBLE  -DCOMPLEX -UDOUBLE -DCOMPLEX -UTRANS -DCONJ -DXCONJ ../kernel/power/cgemv_n.c -o ../cgcc-8_cgemv.S && cd ..
```

**Comment by martin-frbg** [comment]
The change in gcc behaviour could be related to this patch https://www.mail-archive.com/gcc-patches@gcc.gnu.org/msg182366.html

**Comment by quickwritereader** [comment]
I see.  
On power8 and  power9 unaligned load store vsx pairs preferred. But they went on with old instructions.   
On power9 there are even new load store pairs with immediate offset support but big endian.
It's hard to track pairs and endianess. if you mixes pairs you can end swapped answers. 
  
So should we provide some hint or fix?   

I think it is possible to check address and compute missed result for blas1 . 
And the same could be applied to cgemv. 
Other option is to switch to inline assembly for loads.
Ironically I wrote  vector code to make porting easier. 
```
gcc-8 ,incorrectly, using aligned  

memory vector codes over  0x....8
 
address that 8byte aligned.
```

**Comment by martin-frbg** [comment]
Not sure I follow you, but wouldn't any "check address and compute missed result" be more expensive than disabling over-enthusiastic optimization ? (As I understand it, there is no simple way in C to require a specific alignment like the .align or .p2align does in assembly)

**Comment by quickwritereader** [comment]
yes  I still have to add align options to internal buffers and  swap_mask as default always choose data type size. It still can fail tests as passed cgemv matrix pointer by user is default aligned.  


> ~Possible solution. `static const char __attribute__((aligned(16))) swap_mask_arr[]={,,}` with pure assembly I would use constant loading, anyway~
> Alignment did not work. I will recheck ,later.

**Comment by quickwritereader** [comment]
cgemv_n cgemv_t caxpy and cdot were re-written explicitly with vec_vsx_ld so that GCC used new load/store instructions.

**Comment by martin-frbg** [comment]
Fixed by quickwritereader's PR #2243
